src/backend/distributed/utils/background_jobs.c

```diff
@@ -446,22 +473,40 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
 		char *databaseName = get_database_name(MyDatabaseId);
 		char *userName = GetUserNameFromId(task->owner, false);
 
-		MemoryContextSwitchTo(oldContext);
-
+		MemoryContextSwitchTo(TopTransactionContext);
 		PopActiveSnapshot();
 		CommitTransactionCommand();
+		MemoryContextSwitchTo(backgroundTaskContext);
 
-		MemoryContextSwitchTo(perTaskContext);
-
-		/*
-		 * The background worker needs to be started outside of the transaction, otherwise
-		 * it will complain about leaking shared memory segments used, among other things,
-		 * to communicate the output of the backend.
-		 */
+		BackgroundWorkerHandle *handle = NULL;
 		dsm_segment *seg = NULL;
-		BackgroundWorkerHandle *handle =
-			StartCitusBackgroundTaskExecuter(databaseName, userName, task->command,
-											 task->taskid, &seg);
+		StringInfo message = NULL;
+
+		bool handleEntryFound = false;
+		BackgroundExecutorHashEntry *handleEntry =
+			hash_search(backgroundExecutorHandles, (void *) &task->taskid, HASH_ENTER,
+						&handleEntryFound);
+		if (!handleEntryFound)
```

Comment:
Do we plan to have a configurable upper bound on the number of allowed concurrent executions? Probably yes, something like the executor pool size would be neat

Reply:
Note that, we are also discussing how to make shard moves parallel. But then, that decision is probably should be given by `rebalance_table_shards()`, and this bg worker should just pick the next avaliable task(s).

I'm planning to discuss this with @thanodnl and @JelteF today, so maybe decision coming from there could impact how the parallel bg worker works.

For now, I think you can focus on making things work in bg worker in parallel, what we plan to do is probably mostly a follow up task on top of this work anyway.

Reply:
Whenever our monitoring bg detects a new ready task, we spawn a new task executor bg. I noticed that pooling is done at db level with `max_worker_processes` parameter which can only be set at server startup. With the default configuration, we can spawn 5 task executor bg. (other 3 is used by other Citus bgs) When I increase the  `max_worker_processes` param, we can create more parallel task executor bgs.

Reply:
While max_worker_processes is indeed the hard upper bound, we don't really want to hit it.  So, I agree with onder that it's good to have some upper bound on the task executor bg-workers. A separate GUC for this would be in line with how postgres manages everything that consumes from max_worker_processes.